### PR TITLE
perf(over window): skip remaining affected rows when rank is not changed

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

For example:

```sql
rank() over (
  partition by ...
  order by val
)
```

Rows:

```
 row_id | val | rank
--------|-----|------
 100    | 2   | 1
 101    | 6   | 2
 102    | 7   | 3
 103    | 9   | 4
```

When changing `val` of row `101`, it will potenrially affect row `102` and `103`. Previously, no matter what the `val` is changed to, we will re-compute all outputs from row `101` to the end of the partition (row `103`). After this PR, if the output of row `102` is not changed, we will stop the re-computation.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by stdrc** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18950?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#18950** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18950?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#18943** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18943?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stdrc and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/stream/src/executor/over_window/over_partition.rs [hunk 1]
```diff
@@ -425,6 +425,8 @@ impl<'a, S: StateStore> OverPartition<'a, S> {
         Option<RangeInclusive<StateKey>>,
     )> {
         let input_schema_len = table.get_data_types().len() - calls.len();
+        let rank_funcs_only = calls.iter().all(|call| call.kind.is_rank());
```
Comment (kwannoel) [inline src/stream/src/executor/over_window/over_partition.rs:428]: Seems this is static? We can do this check once when instantiating the executor.

File: src/stream/src/executor/over_window/over_partition.rs [hunk 2]
```diff
@@ -530,12 +536,19 @@ impl<'a, S: StateStore> OverPartition<'a, S> {
...
+                let mut should_continue = true;
+
                 let output = states.slide_no_evict_hint()?;
                 compute_count += 1;
 
                 let old_output = &row.as_inner()[input_schema_len..];
                 if !old_output.is_empty() && old_output == output {
                     same_output_count += 1;
+
+                    if rank_funcs_only && key >= last_delta_key {
```
Comment (kwannoel) [inline src/stream/src/executor/over_window/over_partition.rs:548]: Will the rank always be sorted in ascending order?

Comment (stdrc) [reply src/stream/src/executor/over_window/over_partition.rs:548]: The key is memcmp-encoded, which means the order type is considered.

File: src/stream/src/executor/over_window/over_partition.rs (hunk 1 above)
Comment (stdrc) [reply src/stream/src/executor/over_window/over_partition.rs:428]: Yes. Let me eliminate this reevaluation in a separate PR, cuz we also have other things that are evaluated each time when creating `OverPartition` in `OverPartition::new`. To do this, I need to change some code unrelated to the optimization proposed in this PR.

**Review by kwannoel — APPROVED** [review]

**Comment by stdrc** [comment]
After some loose benchmark based on a custom query on nexmark datagen, I found that this PR doesn't help much, but still better than nothing.

Query (no much rationale, just a quick attempt to imitate a user's query):

```sql
CREATE VIEW v1 AS
SELECT
    foo,
    bar,
    SUM(point) AS points
FROM (
    SELECT
        ((bidder * 2654435761) # (bidder >> 16)) % 10000 as foo,
        auction % 3 as bar,
        price % 10 as point
    FROM bid
)
GROUP BY foo, bar;

CREATE SINK mv AS
SELECT
    foo,
    bar,
    points,
    ROW_NUMBER() OVER (
        PARTITION BY bar
        ORDER BY points DESC, foo
    ) AS r
FROM v1
WITH ( connector = 'blackhole', type = 'append-only', force_append_only = 'true');
```

Before:

row number old 1

After:

row number skip same output

I'm going to merge this first, since no more operations can be done on this PR and the performance improvement looks small.

**Pull request merged**